draft-ietf-mls-protocol.md

```diff
@@ -1518,15 +1518,24 @@ to other members in the group via an UpdatePath message (see {{update-paths}}) .
 All other group members then apply the key material in the UpdatePath to their
 own local tree state to derive the group's now-updated shared secret.
 
-To begin, the generator of the UpdatePath updates its leaf
-KeyPackage and its direct path to the root with new secret values.  The
-HPKE leaf public key within the KeyPackage MUST be derived from a freshly
-generated HPKE secret key to provide post-compromise security.
-
-The generator of the UpdatePath starts by sampling a fresh random value called
-"leaf_secret", and uses the leaf_secret to generate their leaf HPKE key pair
-(see {{key-packages}}) and to seed a sequence of "path secrets", one for each
-ancestor of its leaf. In this setting,
+To begin with, the generator of the UpdatePath updates its leaf's KeyPackage and
+its leaf's _filtered direct path_ with new key pairs. The filtered direct path 
+of a leaf is obtained from the leaf's direct path by removing all nodes whose
+child on the leaf's copath has an empty resolution. Such a removed node does not
+need a key pair, since after blanking it, its resolution consists of a single
+node on the filtered direct path. Using the key pair of the node in the
+resolution is equivalent to using the key pair of the removed node.
+
+The generator of the UpdatePath starts by updating the KeyPackage of its leaf.
+The HPKE leaf public key within the KeyPackage MUST be
+derived from a freshly generated HPKE secret key to provide post-compromise
```

Comment:
I'm not sure if that is what you meant here, but I believe the HPKE key pair is generated from a given (symmetric) secret rather than the public key being generated using the private key.